# nnet: Multinomial Logistic Regression

Since it is a type of regression models, the accuracy of prediction is upto how well the dataset is distributed.

## Data

In [51]:
data(iris) 
# remark: iris dataset is already in order 
#         and numbers of targets are exactly the same

## Sampling and Spliting data into training and testing sets

In [44]:
set.seed(1234)
samp <- c(sample(1:50, 35), sample(51:100, 35), sample(101:150, 35))

# Training
iris.train <- iris[samp,]

# Test
iris.test <- iris[-samp,]

# Training

In [45]:
library(nnet)

In [46]:
model <- multinom(Species ~., data = iris.train)
model

# weights:  18 (10 variable)
initial  value 115.354290 
iter  10 value 10.451001
iter  20 value 0.395906
iter  30 value 0.063643
iter  40 value 0.032814
iter  50 value 0.023376
iter  60 value 0.021200
iter  70 value 0.019040
iter  80 value 0.017101
iter  90 value 0.012218
iter 100 value 0.011383
final  value 0.011383 
stopped after 100 iterations


Call:
multinom(formula = Species ~ ., data = iris.train)

Coefficients:
           (Intercept) Sepal.Length Sepal.Width Petal.Length Petal.Width
versicolor    38.03419    -13.46319   -9.680341     26.05544   -12.04250
virginica    -56.23466    -34.56369  -21.990651     60.49496    43.87704

Residual Deviance: 0.02276601 
AIC: 20.02277 

In [47]:
summary(model)

Call:
multinom(formula = Species ~ ., data = iris.train)

Coefficients:
           (Intercept) Sepal.Length Sepal.Width Petal.Length Petal.Width
versicolor    38.03419    -13.46319   -9.680341     26.05544   -12.04250
virginica    -56.23466    -34.56369  -21.990651     60.49496    43.87704

Std. Errors:
           (Intercept) Sepal.Length Sepal.Width Petal.Length Petal.Width
versicolor    2370.414      1467.14    2161.940     1294.101    3264.332
virginica     2372.283      1469.90    2163.035     1299.654    3265.816

Residual Deviance: 0.02276601 
AIC: 20.02277 

## Prediction

In [48]:
pred <- predict(model, iris.test)
pred

[1] setosa     setosa     setosa     setosa     setosa     setosa    
 [7] setosa     setosa     setosa     setosa     setosa     setosa    
[13] setosa     setosa     setosa     versicolor versicolor versicolor
[19] versicolor versicolor versicolor versicolor virginica  versicolor
[25] versicolor versicolor versicolor versicolor virginica  versicolor
[31] virginica  virginica  virginica  virginica  virginica  virginica 
[37] versicolor virginica  virginica  virginica  virginica  virginica 
[43] virginica  virginica  virginica 
Levels: setosa versicolor virginica

In [49]:
summary(pred)

setosa versicolor  virginica 
        15         14         16

## Accuracy

In [50]:
mean(pred == iris.test$Species)

[1] 0.9333333

The model accuracy has turned out to be 93.33%